Fix for image push to use chunked encoding

z4ce opened the pull request:
By not setting the request size, we can let net/http handle chunking the request. This is important for any registry that is behind Cloudflare or another CDN that limits HTTP POST chunk size.

z4ce commented:
It turns out a few of the tests were relying on a quirk of net/http that if you wrote the number of bytes that you specified in `ContentLength` to the body it would send the request and generate a response. I updated two of the tests to close the writer. I removed a third test `TestPusherErrReset` because it didn't seem to be covering any additional code and was highly reliant on the `Write() # of bytes == Content-Length closes body` behavior. I'm not sure what scenario it was actually testing in any case, since by the time `Content-Length == Sum Of Writes` presumably you'd commit at that point anyway.. meaning you'd never make another write that would trigger whatever this test was trying to demonstrate. 

It also did occur to me that these tests would likely be better if they were using the Status to verify the expected behavior rather than reading the response objects off the internal channel, but I did not attempt that.

z4ce commented:
Would also resolve this issue.. by removing the test #7471

lengrongfu commented:
This issue https://github.com/containerd/containerd/issues/7471 has been resolved, please confirm if it resolved your issue.
@z4ce

z4ce commented:
@lengrongfu the issue with the `TestPusherErrReset`? I'm almost positive it will deadlock with chunking enabled. I can rebase my changes based on your fixes, though

z4ce commented:
@lengrongfu I rebased and updated the new tests to not rely on the the content-length==auto-close behavior.  Let me know if you have any questions.  I'm also on the CNCF slack and at Kubecon if want to discuss online or in person

z4ce commented:
FYI the failing test on this seems to just be a flaky test

akhilerm commented:
@z4ce `TestPusherErrReset` is covering a different behaviour. i.e when the registry responds with a redirect, we will have to reset and send the body again. This is seen in case of uploading a mulitplatform image index to quay. Ref: #5978 . 
This test has been added to check the fix for the above mentioned issue.

z4ce commented:
@akhilerm  I think the other tests are handling this scenario now. I tried to understand what code this was covering so I ran this:
```
git reset --hard origin/main
HEAD is now at c21d1baa8 Merge pull request #7626 from akhilerm/update-cherry-pick-process
─ /opt/homebrew/bin/go test -timeout 30s -coverprofile=/Users/ianzink/with-errreset.cover github.com/containerd/containerd/remotes/docker
ok  	github.com/containerd/containerd/remotes/docker	2.500s	coverage: 62.6% of statements
╰─ vi pusher_test.go # edit file to remove the TestPusherErrReset test
╰─ /opt/homebrew/bin/go test -timeout 30s -coverprofile=/Users/ianzink/no-errreset.cover github.com/containerd/containerd/remotes/docker
ok  	github.com/containerd/containerd/remotes/docker	2.441s	coverage: 62.6% of statements
╰─ diff <(sort /Users/ianzink/no-errreset.cover)  <(sort /Users/ianzink/with-errreset.cover)
╰─
```
This leads me to believe that the other tests are covering each scenario sufficiently. If that's not the case, can you help me understand what branch of the code we need to make sure is tested?

akhilerm commented:
I am not sure about the code coverage part. There may be overlap with other tests in `content/helpers_test.go`, that checks the Copy method. This is specifically for testing the push() method when an `ErrReset` occurs.

z4ce commented:
I'm pretty sure this test is not testing what it hopes to test and that's why it's not covering any code.  I think it is important to note that net/http will never attempt to write anything until either Content-Length bytes is written or the pipe is closed.

Let's go through it a bit...we're going to overwrite the put handler to the first time it's called to return an http header that says it's timed out.. ok..
```go
	reg.putHandlerFunc = func() func(w http.ResponseWriter, r *http.Request) bool {
		// sets whether the request should timeout so that a reset error can occur and
		// request will be retried
		shouldTimeout := true
		return func(w http.ResponseWriter, r *http.Request) bool {
			if shouldTimeout {
				shouldTimeout = !shouldTimeout
				w.WriteHeader(http.StatusRequestTimeout)
				return true
			}
			return false
		}
	}()
```
So now we get to the part where we actually write an image
```go
	w, err := p.push(context.Background(), desc, remotes.MakeRefKey(context.Background(), desc), false)
	assert.Equal(t, err, nil, "no error should be there")

	w.Write(ct)
```
Because we are writing the full content length, net/http is going to implicitly make the request. Doesn't seem like we should rely on that and should actually Close the pipe like a real user would after writing all the bytes.

```go
	pw, _ := w.(*pushWriter)

	select {
	case p := <-pw.pipeC:
		p.Write(ct)
	case e := <-pw.errC:
		assert.Failf(t, "error: %v while retrying request", e.Error())
	}

	select {
	case resp := <-pw.respC:
		assert.Equalf(t, resp.StatusCode, http.StatusCreated,
			"201 should be the response code when uploading new content")
	case <-pw.errC:
		assert.Fail(t, "should not give error")
	}
```
So now the real part of the test

1.  we're casting the writer to the internal writer implementation...
2.  Then reading off an internal channel.. which really should have been closed by the caller...
3.  Then we write the channel again.. which no caller will ever do. If anyone writes Content-Length bytes, then they should call close and never write more bytes.
4.  Then we make sure if we write Content-Length bytes AGAIN that another request is made. 

In reality, no caller will ever call Write again which will trigger the second call... likely there is a test already here that tests the scenario where the caller in fact does call close. That is is the `TestPusherErrClosedRetry` test.. which tests the same code as the `TestPusherErrReset` (which can be verified using code coverage).

dmcgowan commented:
> Because we are writing the full content length, net/http is going to implicitly make the request. Doesn't seem like we should rely on that and should actually Close the pipe like a real user would after writing all the bytes.
> ...(which can be verified using code coverage).

This test is just ensuring that `setPipe` is called twice. Since the call is inside the body func called by the `net/http` library, it makes sense to verify the behavior that the reset logic is expecting. Code coverage doesn't really help us here since there we are looking for the same section called again, tested at a higher level it should ensure the err reset is handled properly.

That test could possibly be improved so that it wasn't relying so much on when the http library is sending the request. Normally the API for the content API is Write followed by Commit, rather than Write followed by Close. In this case, using that API to call Commit would end up triggering the read from the internal go routine. When the test was first written the commit logic I think still had a bug though, that test may be able to be refactored to just call `Write`, then `Commit` (which should return `content.ErrReset`, followed by a successful Write+Commit (or Write+Close). The would get rid of the awkward casting to an internal type and just use the exposed interfaces.

z4ce commented:
Thanks, @dmcgowan . Per our conversation on CNCF slack, I updated the test. I believe this covers the situation the test was designed for, but let me know.

SebastianSchildt commented:
Is this PR still active? It seems to me at least the issue still persists

z4ce commented:
It looks like some of my fixes for tests made it into the main through other PRs. I'll rebase this PR, and hopefully eventually it will be pulled. I know there some concern that maybe we don't always want to allow chunking. IMHO leaving it to `net/http` makes the most sense (like moby does), but if the core maintainers want to give me a magic size value for chunkink/non-chunking we can do that too.. cc @dmcgowan

z4ce commented:
The PR is rebased with no conflicts now cc @dmcgowan

z4ce commented:
@dmcgowan is that windows test holding this up? I don't think it is related seems to be a broken test

z4ce commented:
@dmcgowan anything I can do to help get this PR merged?

DevSpork commented:
Hey I'm facing the same issue is it possible to get this merged? I seems as all checks pass.

nopjar commented:
I am facing the exact same issue! Would definitely appreciate this pr to get merged :)

dmcgowan commented:
The discussion this one needs might be better in the OCI group rather than here. The OCI spec defines the chunked uploading technique using `PATCH`, as mentioned in #3533. However, I'm not sure in practice that is the best solution or better than using HTTP chunked body. The distribution specification today explicitly defines the use of content length for the single upload and PATCH for the chunked (https://github.com/opencontainers/distribution-spec/blob/main/spec.md#push). We could probably suggest language there mentioning the content length as optional. Since push is mostly functionality done by client importing the libraries, would also be more interested in their (buildkit, nerdctl, or any other project seeing this issue) take on this than my own.

z4ce commented:
Thanks @dmcgowan. This helps a lot. Interesting.. it seems like Docker/moby always uses the `PATCH` method of push.. and during the PATCH it does not specify the `Content-Length` header, which from what I can tell is technically against the specification you provided.

Here is a screen cap of how docker does it (uses POST+PATCH+PUT):
<img width="1193" alt="docker push" src="https://user-images.githubusercontent.com/4860914/234388361-ea40ddf9-91e0-4f15-8a26-3d2cdcf7cf7c.png">

And this is what containerd looks like (without this patch) it uses the POST+PUT method of upload:
<img width="1196" alt="containerd push" src="https://user-images.githubusercontent.com/4860914/234400536-9926f897-a20d-497f-8c8a-a5882208c1e4.png">


With this patch it looks like:
<img width="1182" alt="containerd with this patch pushing" src="https://user-images.githubusercontent.com/4860914/234403744-5dfcec78-d918-4676-86d9-be616c977a42.png">

z4ce commented:
My inclination would be to ask the specification to be changed to at least reflect the current behavior that docker follows to be included in the specification, and perhaps additionally to allow `Content-Length` to be omitted in the `PUT` method to allow atomic, chunked push. 

In any case, my patch here would need to be updated to send the correct `Content-Length` on the POST request.

If the specification was updated, would you be amenable to pulling a change that would meet that requirement?

dmcgowan commented:
@z4ce I think the best long term solution is to use PATCH for larger blobs and break up the upload into individual PATCH requests of pre-defined static size. The advantage of PATCH is for large uploads, individual requests could be retried multiple times without restarting the entire upload.

dmcgowan commented:
> This is important for any registry that is behind Cloudflare or another CDN that limits HTTP POST chunk size.

What is the limit on some of these CDNs, is it always defined in their documentation?

dmcgowan commented:
Regardless of the breakdown between PATCH and PUT, I think you are right that chunked encoding could be used there. The range header should also be included in the request, were those also missing in the first PATCH example?

> FWIW the HTTP chunked encoding uploads should be retryable and resumable as well.

Are they though in practice and is it possible with Go? Besides fix this specific issue with the chunked encoding, we still have a lot of room to improve the reliability of large layer pushes.

imeoer commented:
https://github.com/containerd/containerd/pull/9405 It's a basic implementation of blob chunked push
Please help to see if it makes sense, thanks! cc @dmcgowan @z4ce

DrJume commented:
It would be good to check if the issue still exists.

olexiyb commented:
I have used cloudflare tunnel to workaround this issue, but cloudflare started to apply the same limits on it and I started to get 413 error on docker push to self hosted goharbor. So we definetely need some solution
